ando/AnDOChecker.py

```diff
@@ -0,0 +1,209 @@
+                file_res.extend(([re.compile(x).search(current_file) is None for x in
+                                  get_list_of_rules_with(rules, Rs.AUTHORIZED_METADATA_EXTENSIONS)]))
+
+            for rules in currentdepth_rules['authorized_data_files']:
+                file_res.extend([re.compile(x).search(current_file) is None for x in
+                                 get_list_of_rules_with(rules, Rs.AUTHORIZED_DATA_EXTENSIONS)])
+                # if none of the authorized rules is respected, raise an error
+
+            if all(file_res):
+                error_list.append("Naming rule not respected for this file : {}".format(current_file))
+        ###
+        # 4. check whether the "mandatory files" are actually present at this level!
+        ###
+        if len(currentdepth_rules['mandatory_files']) > 0:
+            # loop over rules, each rule corresponding to one mandatory file
+            for current_mandatoryfolder_rule in (currentdepth_rules['mandatory_files']):
+                list_of_mandatory_folders = (get_list_of_rules_with
+                                             (current_mandatoryfolder_rule, current_mandatoryfolder_rule[1]))
+                for each_file in list_of_mandatory_folders:
+                    file_res = [re.compile(each_file).search(f) is None for f in files]
+                    if all(file_res):
+                        error_list.append("Mandatory file not found for this rule : {}".
+                                          format(current_mandatoryfolder_rule))
+
+    # if there are no errors, the data set is valid!
+    valid = len(error_list) == 0
+
+    return valid, error_list
+
+
+def get_list_of_rules_with(file_name_regexp, file_ext_regexp=None):
+    """
+    Function that create rules base on the regexp in rulesStructured.py
+
+    Parameters
+    ----------
+
+        file_name_regexp: list
+            list of filenames
+        file_ext_regexp: list
+            list of extensions
+
+    Returns
+    -------
+        list_of_rules : list
+            Concatenate every extension with is respective file
+
+    Examples
+    -------
+    if :
+    file_name_regexp = [
+        ['sub-([a-zA-Z0-9]+)_ses-([a-zA-Z0-9]+)([\\w\\-]*)_ephys'],
+        ['sub-([a-zA-Z0-9]+)_ses-([a-zA-Z0-9]+)([\\w\\-]*)_channels'],
+        ]
+
+    and :
+    file_ext_regexp = ['.tsv', '.json']
+
+    then :
+    list_of_rules = [
+        'sub-([a-zA-Z0-9]+)_ses-([a-zA-Z0-9]+)([\\w\\-]*)_ephys.tsv',
+        'sub-([a-zA-Z0-9]+)_ses-([a-zA-Z0-9]+)([\\w\\-]*)_ephys.json',
+        'sub-([a-zA-Z0-9]+)_ses-([a-zA-Z0-9]+)([\\w\\-]*)_channels.tsv',
+        'sub-([a-zA-Z0-9]+)_ses-([a-zA-Z0-9]+)([\\w\\-]*)_channels.json',
+        ]
+    """
+    list_of_rules = list()
+    if file_ext_regexp is None:
+        list_of_rules = file_name_regexp
+    else:
+        if len(file_name_regexp) == 2:
+            for filename in file_name_regexp[:-1]:
```

Comment:
So the file_name_regexp is not just a list of filenames, but can also contain the potential extensions. Shouldn't then the extension be used to build the rule itself instead of the file_ext_regexp?

Reply:
Yes you're right I should use the last part of the  list with the list of extension to build the rules.But it means that we are not using the 2nd parameter given in argument any more

Reply:
I guess then we can actually remove it...

Reply:
Makes also the docstring shorter ;)

Reply:
I can't remove it because we are using it when there is no extension in the first list

Reply:
Ah, because that would require adding the authorized metadata extensions to all the metadata files in the rules... 
Which might be the conceptually cleaner solution as then the complete definition of all regexes would be in the rules file and no additional assumptions (like metadata file should have metadata information) would be coded in the validator. But also it would make the rule definition more complex again. What do you think?

Reply:
Yes so , I've removed the 2nd argument of the function now all filenames and extension are in the same argument as : 
```[['sub-([a-zA-Z0-9]+)_ses-([a-zA-Z0-9]+)([\\w\\-]*)_ephys',AUTHORIZED_DATA_EXTENSIONS]]```
with ```AUTHORIZED_METADATA_EXTENSIONS = [f'({"|".join([".tsv",".json"])})']```